osd: Support for OMAPs in EC Pools - Classic

MattyWilliams22 opened the pull request:
### This PR prepares the code for enabling support for omap operations in EC pools

Fixes: https://tracker.ceph.com/issues/74188

**ECOmapJournal:**
An in-memory journal to maintain OMAP consistency for Erasure Coded pools.

Background:
Unlike Replicated pools which use a "roll-forward" recovery model, EC pools rely on a
"roll-back" mechanism for interrupted updates. To support efficient rollback without
incurring expensive read-before-write operations to generate rollback info, EC OMAP
updates are initially recorded only in the PG Log. They are applied to the underlying
ObjectStore only after the transaction is fully committed on all shards.

The Problem:
This deferred application creates a latency window where the authoritative state exists
in the PG Log but the ObjectStore contains stale data. Reads served during this window
would return incorrect results.

The Solution:
An ECOmapJournal tracks these "log-only" updates in memory. When an OMAP read occurs,
the backend fetches the base state from the ObjectStore and supplements it with the
updates stored in this journal. This ensures clients always receive the most up-to-date
result, merging the persistent state with the in-flight log state.

Unit tests for the journal have also been added.

**"supports_omap" Pool Flag:**
Added a pool flag to determine whether a pool supports omap operations.
This pool flag is currently always enabled in Replicated pools and always disabled in EC pools.
This flag will be used to enable EC Omap functionality in EC pools in PR 3.

**Omap Read Operation Wrappers:**
In PGBackend, we have introduced abstract omap read operations methods.
In ReplicatedBackend, these methods maintain existing behaviour.
In ECSwitch: 
- If optimisations are not enabled, these methods maintain existing behaviour.
- If optimisations are enabled, these methods combine the updates from the ec_omap_journal with the contents of the object store.

**ObjectModDesc::Visitor:**
Added EC_OMAP ModID and a function to encode omap updates.

**Trimmer:**
Added function to commit omap updates and remove the corresponding entry from the journal.


#### Recovery

- Add omap fields to read_request_t, read_result_t, ECSubRead and ECSubReadReply. 
- Read and write omap header and entries if !omap_complete.
- Require omap_complete to finish recovery.

**Integrating EC Omaps:**
The code that allows for Omap operations in EC pools, including rollbacks and recovery, has been connected up with the rest of Ceph.
This includes:
- Changing recovery reads to include journal updates.
- Clearing the journal for a new epoch.
- Setting the omap_complete boolean correctly before recovery in an EC pool.
- Encoding omap updates and adding entries to the journal.
- Decoding omap updates, applying updates to object store, then removing updates from the journal.
- Changing the supports_omap function to use the pool flag. This function also returns true if is_replicated, this ensures that omap operations are handled during an upgrade, before the pool flag has been set.
- Changing omap read operations in PrimaryLogPG to use PGBackend functions (which include journal updates for optimised_ec).

**Integration Tests:**
- A test suite has been added allowing for tests to be run against ec pools with optimisations enabled.
- We have tested all of the omap object operations.
- We have tested omap recovery when the primary osd in changed.
- We have tested that object recovery still works with no omap present.
- We have tested that omap recovery works when multiple reads are required to retrieve all of the omap data.
- These integrations tests currently skip the tests with FastEC pools.

**Documentation:**
Documentation updated to reflect the fact that omap in now supported in EC pools, but disabled by default.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

MattyWilliams22 committed:
osd: Add ECOmapJournal class and relocate OmapUpdateType enum class

The ECOmapJournal will be used to store omap updates (in ec pools with optimisations enabled) which have not yet been committed to the object store.
Added unit tests for this class.
Promoted OmapUpdateType to osd_types.h so that it can be shared to multiple files without circular dependencies.

Signed-off-by: Matty Williams <Matty.Williams@ibm.com>

MattyWilliams22 committed:
osd: Introduce functions required for EC OMAP support

Introduced a "supports_omap" pool flag which is always enabled for Replicated pools and currently always disabled for EC pools.
Introduced wrappers around omap read operations in PGBackend to include updates from the journal in EC pools with optimisations enabled.
Introduced a function for encoding an EC_OMAP operation in the ObjectModDesc::Visitor class and a function for committing an operation in the Trimmer struct.

Signed-off-by: Matty Williams <Matty.Williams@ibm.com>

MattyWilliams22 committed:
test: Add integration tests for EC Omap operations and recovery

Assisted-by: Bob
Used for writing tests following the pattern of existing tests.

Fixes: https://tracker.ceph.com/issues/74188
Signed-off-by: Matty Williams <Matty.Williams@ibm.com>

MattyWilliams22 commented:
jenkins test make check arm64

MattyWilliams22 commented:
jenkins test windows

MattyWilliams22 commented:
jenkins test make check arm64

MattyWilliams22 commented:
jenkins test windows

MattyWilliams22 committed:
doc: Update documentation to reflect new functionality

https://tracker.ceph.com/issues/74188
Signed-off-by: Matty Williams <Matty.Williams@ibm.com>

bill-scales commented on src/mon/OSDMonitor.cc:
```diff
@@ -6502,7 +6503,11 @@ bool OSDMonitor::preprocess_command(MonOpRequestRef op)
-	}
+	  case SUPPORTS_OMAP:
+	    f->dump_bool("supports_omap",
+                         p->has_flag(pg_pool_t::FLAG_OMAP));
```
I think you should use the support_omap query instead so that replicated pools report true even before require_osd_release has been increased

bill-scales commented on src/mon/OSDMonitor.cc:
```diff
@@ -8271,6 +8281,9 @@ int OSDMonitor::prepare_new_pool(string& name,
   if (crimson) {
     pi->set_flag(pg_pool_t::FLAG_CRIMSON);
   }
+  if (pool_type == pg_pool_t::TYPE_REPLICATED) {
```
nit, technically you should check require_osd_release >= umbrella here to avoid setting the flag when there is still a possibility that some OSDs are running pre-umbrella code or could be downgraded to pre-umbrella code. I think in practice earlier code will ignore this flag

MattyWilliams22 committed:
doc: Write design document to explain the reasoning behind implementing this feature

Assisted-by: Bob
Used to create the first draft of the design document.

https://tracker.ceph.com/issues/74187
Signed-off-by: Matty Williams <Matty.Williams@ibm.com>

MattyWilliams22 committed:
osd: Allow for recovery of OMAP header and entries in EC pools

Add omap fields to read_request_t, read_result_t, ECSubRead and ECSubReadReply.
Read and write omap header and entries if !omap_complete.
Require omap_complete to finish recovery.

Fixes: https://tracker.ceph.com/issues/74244
Signed-off-by: Matty Williams <Matty.Williams@ibm.com>

MattyWilliams22 commented:
jenkins test submodules

bill-scales commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -10117,11 +10181,15 @@ void PrimaryLogPG::finish_copyfrom(CopyFromCallback *cb)
 {
   OpContext *ctx = cb->ctx;
   dout(20) << "finish_copyfrom on " << ctx->obs->oi.soid << dendl;
+  dout(20) << __func__ << " COPY_FROM operation completing on OSD "
```
delete this debug

bill-scales commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9513,10 +9561,13 @@ void PrimaryLogPG::start_copy(CopyCallback *cb, ObjectContextRef obc,
+           << " flags " << flags
+           << (mirror_snapset ? " mirror_snapset" : "")
+           << dendl;
+  dout(20) << __func__ << " initiating copy operation on OSD "
```
delete this debug

bill-scales commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9726,6 +9777,15 @@ void PrimaryLogPG::process_copy_chunk(hobject_t oid, ceph_tid_t tid, int r)
     return;
   }
 
+  dout(20) << __func__ << " received omap_header length: "
```
delete these 2 debug statements

bill-scales commented on src/osd/ECBackend.cc:
```diff
@@ -94,6 +94,8 @@ ECBackend::ECBackend(
   ceph_assert((ec_impl->get_data_chunk_count() *
     ec_impl->get_chunk_size(stripe_width)) == stripe_width);
+
+  ec_omap_journal.set_dpp(get_parent()->get_dpp());
```
nit: Might be nicer to pass in dpp in the Constructor and then remove the if(dpp) logic in ec_omap_journal

MattyWilliams22 replied:
ec_omap_journal is initialised in ECCommon, which does not have a constructor. Is it worth creating a constructor for ECCommon for this change? Is there a way to pass dpp into the ECOmapJournal constructor without creating a constructor for ECCommon?

bill-scales replied:
I think you would need to add a constructor for ECCommon as well

bill-scales commented on src/osd/PeeringState.cc:
```diff
@@ -4741,7 +4741,11 @@ void PeeringState::append_log(
       * object is deleted before we can _merge_object_divergent_entries().
       */
-    pg_log.skip_rollforward(&info, handler.get());
+    psdout(20) << __func__ << " skip_rollforward starting, can_rollback_to="
```
delete 2 x debug statements

bill-scales commented on src/osd/PeeringState.cc:
```diff
@@ -4761,8 +4765,11 @@ void PeeringState::append_log(
+      psdout(20) << __func__
+             << ": rolling forward past last_backfill, soid="
+             << p->soid << " last_backfill=" << info.last_backfill
+             << " entry=" << *p << dendl;
```
This change should be in a separate commit with a description about what the change is achieving. Its a tiny change but with big consequences.

The old code never got executed because EC+Replica always skipped sending the write message in these conditions. However this then meant that writes to backfilling shards (not in acting set) that were below the last_backfill watermark did not get rolled forward here and could get skipped by a later update calling skip_rollforward.

Prior to this PR skipping a roll forward could forget to clean up generation'd objects (and hence leak storage) but would otherwise not have any ill effects. The code doesn't do a good job of policing that these objects eventually get deleted, so this would go unnoticed. This PR makes omap updates when the log entry is rolled forward, so skipping it causes omap updates to be dropped and a data integrity issue.

As the comments above say, it is safe to roll forward updates to shards that are not in the acting set as soon as they are applied because these shards are not used by Peering when determining the authoritative log and they do not need to be read for recovery. The worst consequence of rolling forward updates on these shards early is that if the update later gets rolled back the object will be marked as missing on this shard and will generate a little extra recovery work.

MattyWilliams22 replied:
I have created a draft PR for this: https://github.com/ceph/ceph/pull/68888

I will remove that code from this PR and remove the debug statements to address your previous comment.

aainscow commented on src/crimson/osd/pg.h:
```diff
@@ -619,6 +619,9 @@ class PG
       // TODO
     }
+    void trim_after_remove(const pg_log_entry_t &entry) override {
+      // TODO
```
Do you know why we have TODO here?  Looks to me like the above were missed... and maybe we shouldn't copy it.

MattyWilliams22 replied:
To match the classic OSD implementation, this would have to call a method with the same name in PGBackend, which would use the TrimmerPostRemove. In crimson, these methods don't exist in pg_backend, and the Trimmers also don't exist yet.

bill-scales replied:
The issue is that PGBackend is for classic OSD code, there is an alternative pg_backend for Crimson. I think the functions in PGBackend that are called by PGLogEntryHandler for rollforward/back/etc would all work fine for Crimson as well, the challenge is refactoring the code to separate them from the rest of PGBackend. Maybe we need a PGCommon?

@rzarzynski I assume this is somewhere on your TODO list for Crimson EC. Let us know if you want to do any more than just leave a TODO here.

aainscow commented on src/osd/ECBackend.cc:
```diff
@@ -570,8 +571,132 @@ void ECBackend::handle_sub_read(
+              << ": " << e.what() << dendl;
+      get_parent()->clog_warn() << "corrupt object info attribute for " << *i;
+      reply->errors[*i] = -EIO;
+    }
```
continue?

aainscow commented on src/osd/ECBackend.cc:
```diff
@@ -1377,3 +1591,257 @@ int ECBackend::be_deep_scrub(
+    return false;
+  }
+
+  // Find range that comes after this key
```
nit: Unnecessary whitespace change.

I know its removing trailing space, but that change doesn't belong here.

MattyWilliams22 replied:
I'm not sure what whitespace change this is referring to.

aainscow replied:
Again - I am not quite sure where this came from.  Ihave been experimenting with a the github tool in CLION - I wonder if there is changes...

aainscow commented on src/osd/ECOmapJournal.h:
```diff
@@ -0,0 +1,153 @@
+  using const_iterator = std::list<ECOmapJournalEntry>::const_iterator;
+ private:
+  // Unprocessed journal entries
+  std::map<hobject_t, std::list<ECOmapJournalEntry>> entries;
```
nit: Unnecessary whitespace change

MattyWilliams22 replied:
This file did not exist before this commit. What are you referring to here?

aainscow replied:
Not sure.... Its only whitespace anyway, so ignore.

aainscow commented on src/osd/ECTransaction.cc:
```diff
@@ -579,15 +790,32 @@ ECTransaction::Generate::Generate(PGTransaction &t,
-      object_info_t oi_updates(*(op.attr_updates[OI_ATTR]));
-      clear_whiteout = !oi_updates.test_flag(object_info_t::FLAG_WHITEOUT);
+  if (op.attr_updates.count(OI_ATTR)) {
+    bufferlist &bl = op.attr_updates.find(OI_ATTR)->second.value();
```
Why have you change contains -> count?  

contains is more modern and permitted in Ceph.

aainscow commented on src/osd/PeeringState.cc:
```diff
@@ -4826,7 +4826,7 @@ void PeeringState::append_log(
     if (transaction_applied &&
-	p->soid > info.last_backfill) {
+    p->soid > info.last_backfill) {
       pg_log.roll_forward(&info, handler.get());
```
nit: Unnecessary whitespace change

aainscow commented on src/osd/PGLog.h:
```diff
@@ -210,13 +212,13 @@ struct PGLog : DoutPrefixProvider {
-	f(*rollback_info_trimmed_to_riter, previous_version);
-	previous_version = rollback_info_trimmed_to_riter->version;
+        --rollback_info_trimmed_to_riter;
+        if (rollback_info_trimmed_to_riter->version > rollback_info_trimmed_to) {
```
nit: Unnecessary whitespace change

aainscow commented on src/osd/PGLog.h:
```diff
@@ -269,32 +271,32 @@ struct PGLog : DoutPrefixProvider {
-	  h->partial_write(info, previous_version, entry);
-	});
+        to,
+        [&](pg_log_entry_t &entry, eversion_t previous_version) {
```
nit: Unnecessary whitespace change (and linex below)

aainscow commented on src/osd/ECOmapJournal.h:
```diff
@@ -0,0 +1,153 @@
+  void process_entries(const hobject_t &hoid);
+  bool remove_processed_entry(const hobject_t &hoid, const ECOmapJournalEntry &entry);
+  bool remove_processed_entry_by_version(const hobject_t &hoid, const eversion_t version);
+  UpdateMapType get_key_map(const hobject_t &hoid);
```
nit: Should be const?

aainscow commented on src/osd/ECOmapJournal.cc:
```diff
@@ -0,0 +1,596 @@
+std::pair<gen_t, bool> ECOmapJournal::get_generation(const hobject_t &hoid) const {
+  if (const auto it = object_state_map.find(hoid); it != object_state_map.end()) {
+    if (const auto& versions = it->second; !versions.empty()) {
+      gen_t gen = versions.begin()->first;
```
niti nit: might look neater as:

const auto& [gen, lost] = *versions.begin();

aainscow commented on src/osd/ECOmapJournal.cc:
```diff
@@ -0,0 +1,596 @@
+            !prev->second || *prev->second >= start) {
+            start = prev->first;
+            if (!end) {
+
```
This looks like a bug - although I don't think it is. Could at least do with a comment.

aainscow commented on src/osd/ECOmapJournal.cc:
```diff
@@ -0,0 +1,596 @@
+  
+  // Get the total number of versions for this object after the operation
+  size_t total_versions = 0;
+  if (const auto it = object_state_map.find(hoid); it != object_state_map.end()) {
```
There are ways of re-arranging this code to avoid the second map lookup, which would help performance. 

e.g. 

auto [it, inserted] = object_state_map.try_emplace(hoid, std::map<version_t, bool>{});

if (!inserted) {
   it->second.insert({version, lost_delete});
}

total_versions = it->second.size();

There is also:

auto& versions = object_state_map[hoid];
versions.insert({version, lost_delete});

size_t total_versions = versions.size();

Which is much neater, but relies on there being a default constructor... so might not even.

aainscow reviewed (approved):
I don't think I have any serious conerns... I agree with Bill's concerns that the dpp could be handled better and simplified. 

I have spent quite a bit of time looking through this code, helping with debugging. I was able to easily understand what was happening. 

I have approved - as I only have nits.

aainscow reviewed (changes requested):
Actually - I think you were using Bob. Even if most of the code was being hand-written, you should still acknowledge the AI assistance and how much you used it.

MattyWilliams22 commented:
jenkins test make check arm64

bill-scales reviewed (approved):
No further issues beyond Alex's comments

MattyWilliams22 committed:
osd: Hook up omap operations in EC pools

Add pool flag to determine if omap operations are supported in a pool.
- Currently disabled in EC pools (will later be enabled for Fast EC pools)
Require all osds to have umbrella or later release version to enable pool flag.
Change recovery reads to use journal updates.
Clear the journal for a new epoch.
Set omap_complete accurately before recovery.
Encode omap updates and add entry to journal.
Decode omap updates, apply updates to object store, then remove from journal.
Change omap reads in PrimaryLogPG to use PGBackend functions, including omap updates from journal.

Assisted-by: Bob
Used for debugging and copying patterns (e.g. implementing REPLACE type to match MODIFY).

Fixes: https://tracker.ceph.com/issues/74188
Signed-off-by: Matty Williams <Matty.Williams@ibm.com>

aainscow reviewed (approved):
LGTM

rzarzynski commented on src/osd/ECOmapJournal.h:
```diff
@@ -0,0 +1,154 @@
+ * in the PG Log but the ObjectStore contains stale data. Reads served during this window
+ * would return incorrect results.
+ *
+ * The Solution:
+ * An ECOmapJournal tracks these "log-only" updates in memory. When an OMAP read occurs,
+ * the backend fetches the base state from the ObjectStore and supplements it with the
+ * updates stored in this journal. This ensures clients always receive the most up-to-date
+ * result, merging the persistent state with the in-flight log state.
```
Makes sense.

rzarzynski commented on src/osd/ECOmapJournal.h:
```diff
@@ -0,0 +1,154 @@
+  std::vector<std::pair<OmapUpdateType, ceph::buffer::list>> omap_updates;
+
+  ECOmapJournalEntry(
+    eversion_t version, bool clear_omap, std::optional<ceph::buffer::list> omap_header,
```
it would be more readable if it were written in the one-param-per-line manner.

rzarzynski commented on src/crimson/osd/ec_backend.cc:
```diff
@@ -45,6 +45,7 @@ ECBackend::ECBackend(pg_shard_t whoami,
 		     DoutPrefixProvider &dpp,
 		     ECListener &eclistener)
   : PGBackend{whoami, coll, shard_services, store_index, dpp},
+    ECCommon(dpp),
```
CC: @Matan-B.

rzarzynski commented on src/osd/PGTransaction.h:
```diff
@@ -137,7 +137,6 @@ class PGTransaction {
 
     std::map<std::string, std::optional<ceph::buffer::list> > attr_updates;
 
-    enum class OmapUpdateType {Remove, Insert, RemoveRange};
```
OK, we're just making it independent and  moving to `osd_types.h`.

rzarzynski commented on src/mon/OSDMonitor.cc:
```diff
@@ -8277,6 +8285,10 @@ int OSDMonitor::prepare_new_pool(string& name,
     pi->set_flag(pg_pool_t::FLAG_CRIMSON);
   }
+  if (pool_type == pg_pool_t::TYPE_REPLICATED
+      && osdmap.require_osd_release >= ceph_release_t::umbrella) {
```
What will happen if – after upgrade to Umbrella BUT without bumping `require_osd_release` up, new pool is created?

batrick replied:
@MattyWilliams22 ??

MattyWilliams22 replied:
Replicated pools always have omap support enabled (this is enforced by adding the `|| is_replicated` to the supports_omap method).
EC pools by default do not have omap support enabled. And they can only have omap support enabled if the require_osd_release is >= umbrella (see the changes to OSDMonitor.cc ~line 9153). 
So, this should not be a problem for Replicated or EC pools.

rzarzynski commented on src/osd/osd_types.h:
```diff
@@ -1811,7 +1815,7 @@ struct pg_pool_t {
 
   bool supports_omap() const {
-    return !(get_type() == TYPE_ERASURE);
+    return has_flag(FLAG_OMAP) || is_replicated();
```
Oh, I see.

rzarzynski commented on src/osd/PGBackend.cc:
```diff
@@ -377,6 +378,80 @@ struct Trimmer : public ObjectModDesc::Visitor {
+            ghobject_t(soid, gen, shard));
+        }
+
+        for (auto &&up: omap_updates) {
+          switch (up.first) {
+            case OmapUpdateType::Remove:
+              t->omap_rmkeys(
+                coll_t(spg),
+                ghobject_t(soid, gen, shard),
+                up.second);
+              break;
+            case OmapUpdateType::Insert:
+              t->omap_setkeys(
+                coll_t(spg),
+                ghobject_t(soid, gen, shard),
+                up.second);
+              break;
+            case OmapUpdateType::RemoveRange:
+              t->omap_rmkeyrange(
+                coll_t(spg),
+                ghobject_t(soid, gen, shard),
+                up.second);
+              break;
+          }
+        }
```
OK, here we're sending the in-memory OMAP updates to object store.

rzarzynski commented on doc/dev/osd_internals/erasure_coding/client_support.rst:
```diff
@@ -0,0 +1,918 @@
+  which are required for Cls operations to function correctly.
+
+These limitations prevented EC pools from being used for many important
+workloads, particularly RBD (RADOS Block Device) which relies heavily on both
+omap and Cls operations.
```
Wasn't RGW the intended audience? RBD isn't so OMAP-intensive.

bill-scales replied:
Although RBD does not make heavy use of omaps, it is the part of the product that suffers most from having to use two pools to make use of EC storage. This is because creating/deleting a block image is a much more common operation than creating/deleting an object store or creating/deleting a file system and both pools need to be specified every time an image is created. There is also a lot more orchestration software (e.g. NVMe G/W, CSI drivers, RBD Mirroring, ...) that create/delete block images. Teaching all of this software to deal with two pools is much harder than fixing EC so that it supports omaps.

This feature may also help simplify CephFS and RGW configurations by requiring less pools to be configured and avoiding misconfiguration issues where a user chooses a Replica pool and EC pool with different amounts of redundancy. However in particular with CephFS there may be other performance reasons why storing a metadata pool on Replica might be advantageous.

anthonyeleven replied:
This.  I’ve seen block clients that can’t use RBD EC because they have no provision for specifying a second pool.  I’m curious why we designed it that way, vs transparently encoding the metadata pool as an attribute of the data pool.  Perhaps since the client would still need the code path this obviates more subtle errors?

With CephFS one absolutely can see issues with slow metadata ops.  I haven’t poked around with RADOS ls to see if meta pools store objects large enough that historical EC padding would be problematic. But so long as EC has a “significant” performance impact we probably want to stick with replication.  Indeed some have advised R4 for CephFS meta pools to get more nines.  

With RGW some multizone deployments share a single index pool, though istr that this precludes either multisite replication or multiple storage classes.

rzarzynski commented on doc/dev/osd_internals/erasure_coding/client_support.rst:
```diff
@@ -0,0 +1,918 @@
+- ``supports_omap = true``
+
+These settings can be configured per-pool. The cluster will enforce that all
+OSDs are at the Umbrella release before allowing omap support to be enabled.
```
🥇 for the documentation!

anthonyeleven replied:
I’ve been encouraging the docs to note when things start or stop working.  With search engines people often end up reading the docs for a release that doesn’t match their cluster.

rzarzynski commented on src/osd/ECMsgTypes.cc:
```diff
@@ -349,6 +376,9 @@ void ECSubReadReply::encode(bufferlist &p_bl,
   encode(errors, p_bl);
+  encode(omap_headers_read, p_bl);
+  encode(omap_entries_read, p_bl);
+  encode(omaps_complete, p_bl);
```
Oops, if the peer has no `SERVER_TENTACLE` in its features bits, we declare the 1st schema version but the new members from 3rd are still appended.

rzarzynski commented on src/osd/ECMsgTypes.cc:
```diff
@@ -219,18 +219,20 @@ void ECSubRead::encode(bufferlist &bl, uint64_t features) const
   }
 
-  ENCODE_START(3, 2, bl);
+  ENCODE_START(4, 2, bl);
   encode(from, bl);
   encode(tid, bl);
   encode(to_read, bl);
   encode(attrs_to_read, bl);
   encode(subchunks, bl);
+  encode(omap_read_from, bl);
+  encode(omap_headers_to_read, bl);
```
ACK.

rzarzynski commented on src/osd/ECMsgTypes.cc:
```diff
@@ -254,6 +256,13 @@ void ECSubRead::decode(bufferlist::const_iterator &bl)
       subchunks[i.first].push_back(make_pair(0, 1));
     }
   }
+  if (struct_v >= 4) {
+    decode(omap_read_from, bl);
+    decode(omap_headers_to_read, bl);
+  } else {
+    omap_read_from.clear();
+    omap_headers_to_read.clear();
+  }
   DECODE_FINISH(bl);
```
ACK.

rzarzynski commented on doc/glossary.rst:
```diff
@@ -352,7 +352,10 @@
                 Ceph cluster. RGW bucket indexes are stored as omaps.
-                Erasure-coded pools cannot store RADOS omap data structures.
+                Erasure-coded pools, by default, cannot store RADOS omap data
+                structures. This functionality can be enabled in erasure-coded
+                pools with optimizations enabled by setting the
+                ``supports_omap`` flag to true.
```
🚀

rzarzynski reviewed (commented):
Basically LGTM **but with a (rather complex) dencoding issue**. Please see the comment.

yuriw commented:
This PR has been added to [wip-yuri10-testing-20260526.155424-main](https://tracker.ceph.com/issues/76813).

rzarzynski commented:
@MattyWilliams22, could you please rebase?

rzarzynski commented on src/osd/ECMsgTypes.cc:
```diff
@@ -349,6 +376,11 @@ void ECSubReadReply::encode(bufferlist &p_bl,
   }
   encode(attrs_read, p_bl);
   encode(errors, p_bl);
+  if (ver >= 3) {
+    encode(omap_headers_read, p_bl);
+    encode(omap_entries_read, p_bl);
+    encode(omaps_complete, p_bl);
+  }
```
LGTM.

batrick commented on doc/cephfs/createfs.rst:
```diff
@@ -155,5 +155,13 @@ If you are storing lots of small files or are frequently modifying files you can
+    ceph osd pool set my_ec_pool supports_omap true
+
+RADOS *OMAP* data structures can only be enabled in Erasure Coded pools that allow ec optimizations.
+Alternatively, you can use a Replicated pool as a metadata pool, while using an Erasure Coded pool as the data pool.
```
I'm not comfortable updating this guidance until this change is mature.

Also: please note this guidance is specific to the **metadata pool**. I'm not aware of a good reason to ever recommend an EC metadata pool even if omap were supported and mature.

batrick commented on doc/dev/osd_internals/erasure_coding/client_support.rst:
```diff
@@ -0,0 +1,918 @@
+By bringing omap and synchronous reads to EC pools, CephFS can store its complex directory
+hierarchies and metadata structures directly alongside file data. This eliminates the
+necessity for a segregated replicated metadata pool, optimizing cluster footprint and
+simplifying capacity planning for file-based workloads.
```
> By bringing omap and synchronous reads to EC pools, CephFS can store its complex directory
> hierarchies and metadata structures directly alongside file data. This eliminates the
> necessity for a segregated replicated metadata pool, optimizing cluster footprint and
> simplifying capacity planning for file-based workloads.

surely this was clearly written by AI? I have no idea eliminating the metadata pool was motivation for this work?

batrick commented on doc/dev/osd_internals/erasure_coding/client_support.rst:
```diff
@@ -0,0 +1,918 @@
+- **EC Omap in Teuthology Tests**: Allow omap operations in EC pools with ceph_test_rados
+- **RBD in Teuthology Tests**: Change all uses of RBD in teuthology to use just an EC pool
+- **RGW in Teuthology Tests**: Change all uses of RGW in teuthology to use just an EC pool
+- **CephFS in Teuthology Tests**: Change all uses of CephFS in teuthology to use just an EC pool
```
??????

batrick reviewed (changes requested):
Needs a rebase.

I'm skeptical about including this in Umbrella. This is very late to be making such a feature available in a new release.

MattyWilliams22 commented:
jenkins test make check arm64

MattyWilliams22 commented:
jenkins test api

MattyWilliams22 commented:
jenkins test windows

rzarzynski commented on doc/rados/operations/erasure-code.rst:
```diff
@@ -197,9 +197,9 @@ should be migrated to BlueStore.
-Erasure-coded pools do not support omap, so to use them with RBD and
-CephFS you must instruct them and their clients to store their data in an EC pool and
-their metadata in a replicated pool. For RBD, this means using the
+RBD and CephFS are not fully supported in erasure-coded pools, instead
+you must instruct them and their clients to store their data in an EC
+pool and their metadata in a replicated pool. For RBD, this means using the
```
@batrick: the doc update is toned down significantly to ensure it cannot be interpreted as a recommendation to use EC OMAP from the beginning.
As `supports_omap` for EC is `false` by default, somebody still needs to opt-in which is fine for me. I'm for pulling this PR into Umbrella.
When the feature gains broader adoption, we can reiterate the documentation update.

ljflores reviewed (commented):
Hey @MattyWilliams22 , there was one issue related to this PR in teuthology:
https://pulpito.ceph.com/yuriw-2026-05-26_20:12:02-rados-wip-yuri10-testing-20260526.155424-main-distro-default-trial/238594
```
2026-05-26T20:56:08.036 INFO:tasks.workunit.client.0.trial179.stderr:set-quota max_objects = 1 for pool 9e55a17a-065c-4819-a308-3519ca17d41e
2026-05-26T20:56:08.052 INFO:tasks.workunit.client.0.trial179.stderr:++ mktemp fooattrXXXXXXX
2026-05-26T20:56:08.053 INFO:tasks.workunit.client.0.trial179.stderr:+ V1=fooattronkeEUX
2026-05-26T20:56:08.053 INFO:tasks.workunit.client.0.trial179.stderr:+ /usr/bin/rados put test_rados_obj fooattronkeEUX -p 9e55a17a-065c-4819-a308-3519ca17d41e
2026-05-26T20:56:08.072 INFO:tasks.workunit.client.0.trial179.stderr:+ /usr/bin/ceph osd dump
2026-05-26T20:56:08.072 INFO:tasks.workunit.client.0.trial179.stderr:+ grep 'full_quota max_objects'
...
2026-05-26T23:54:02.801 INFO:tasks.workunit.client.0.trial179.stderr:Terminated
2026-05-26T23:54:02.801 INFO:tasks.workunit.client.0.trial179.stderr:++ rm -rf /tmp/test_rados_tool.TiVErWsuU2
2026-05-26T23:54:02.803 INFO:tasks.workunit.client.0.trial179.stderr:++ exit
2026-05-26T23:54:02.803 INFO:tasks.workunit.client.0.trial179.stderr:+ rm -rf /tmp/test_rados_tool.TiVErWsuU2
2026-05-26T23:54:02.804 INFO:tasks.workunit.client.0.trial179.stderr:+ exit
2026-05-26T23:54:02.805 DEBUG:teuthology.orchestra.run:got remote process result: 124
2026-05-26T23:54:02.805 INFO:tasks.workunit:Stopping ['rados/test_rados_tool.sh'] on client.0...
2026-05-26T23:54:02.806 DEBUG:teuthology.orchestra.run.trial179:> sudo rm -rf -- /home/ubuntu/cephtest/workunits.list.client.0 /home/ubuntu/cephtest/clone.client.0
2026-05-26T23:54:03.074 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
```
This issue is happening because a new pool flag was introduced, and the "grep" check is highly dependent on specific flag ordering.

I saw that a separate EC PR also adds a new pool flag and addresses this by changing the "ceph osd dump / grep" check to JSON output verified by a "jq" query:
https://github.com/ceph/ceph/pull/66459/changes/06034c676c42c404c2bdc75245fcdf22d86b0111

You may want to consider adding @JonBailey1993's commit to your PR (I'd check with him first though).

Here is the whole PR: https://github.com/ceph/ceph/pull/66459

batrick commented on doc/cephfs/createfs.rst:
```diff
@@ -156,4 +156,3 @@ If you are storing lots of small files or are frequently modifying files you can
     ceph osd pool set my_ec_pool allow_ec_optimizations true
 
 You may not use Erasure Coded pools as CephFS metadata pools, because CephFS metadata is stored using RADOS *OMAP* data structures, which EC pools cannot store.
-
```
nit: whitespace churn

batrick commented on src/mon/MonCommands.h:
```diff
@@ -1170,11 +1170,11 @@ COMMAND("osd pool rename "
 COMMAND("osd pool set "
 	"name=pool,type=CephPoolname "
-	"name=var,type=CephChoices,strings=size|min_size|pg_num|pgp_num|pgp_num_actual|crush_rule|hashpspool|nodelete|nopgchange|nosizechange|write_fadvise_dontneed|noscrub|nodeep-scrub|hit_set_type|hit_set_period|hit_set_count|hit_set_fpp|use_gmt_hitset|target_max_bytes|target_max_objects|cache_target_dirty_ratio|cache_target_dirty_high_ratio|cache_target_full_ratio|cache_min_flush_age|cache_min_evict_ ...
+	"name=var,type=CephChoices,strings=size|min_size|pg_num|pgp_num|pgp_num_actual|crush_rule|hashpspool|nodelete|nopgchange|nosizechange|write_fadvise_dontneed|noscrub|nodeep-scrub|hit_set_type|hit_set_period|hit_set_count|hit_set_fpp|use_gmt_hitset|target_max_bytes|target_max_objects|cache_target_dirty_ratio|cache_target_dirty_high_ratio|cache_target_full_ratio|cache_min_flush_age|cache_min_evict_ ...
```
I would have preferred `ec_supports_omap` but it may be too late to fix this.

bill-scales replied:
We wanted to make it a general property of a pool that code could query. There is a lot of current code that was checking whether the pool type was erasure code to determine whether omap was supported or not. We didn't want to replace all these checks with "is pool replica || is ec_supports_omap set", a "is supports_omap" check seemed cleaner

bill-scales replied:
@MattyWilliams22 - with regards to this https://github.com/ceph/ceph/pull/66459#discussion_r3345526166 can you break the "name=" line in a similar way to other commands such as "fs set" so each string is on a separate line to improve readability and reduce merge issues

batrick replied:
I wrote that comment before I saw how it was enabled for replicated pools automatically. I like this approach.

batrick commented on src/mon/OSDMonitor.cc:
```diff
@@ -12279,6 +12310,18 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
+          dout(10) << __func__ << " replicated pool " << pool_id
+                   << " has OMAP support auto-enabled" << dendl;
+        }
+      }
```
This is a clever place to put this upgrade.

batrick replied:
As clever as it is, this seems to be the conventional place to put it:

https://github.com/ceph/ceph/pull/66459/changes#diff-c44edb74ef654677b989657a8dad6cd87b952012f21fc055c2f531e7ab637b3bR1967

it's also where I expected to see this upgrade.

bill-scales replied:
Agree, that we should follow convention - we don't want to scatter upgrade code in multiple places

batrick commented on doc/dev/osd_internals/erasure_coding/client_support.rst:
```diff
@@ -0,0 +1,914 @@
+Required Settings
+~~~~~~~~~~~~~~~~~
+
+To enable the new features, the following settings are required:
```
```suggestion
To enable the new features, the following OSDMap pool settings are required:
```

batrick reviewed (approved):
This is fine work and I want to echo @rzarzynski that the design doc is high quality.

Well done!

batrick commented on src/mon/MonCommands.h:
```diff
@@ -1170,11 +1170,136 @@ COMMAND("osd pool rename "
+          "|allow_ec_optimizations"
+          "|ec_data_shard_count"
+          "|ec_coding_shard_count"
+          "|supports_omap",
```
Thanks for doing this.

I recommend you sort these lexicographically to discourage appending to the list (for lack of a better mechanism). Appending also generates needless conflicts.

batrick commented on src/mon/MonCommands.h:
```diff
@@ -1170,11 +1170,136 @@ COMMAND("osd pool rename "
+          "|allow_ec_optimizations"
+          "|set_pool_flags"
+          "|unset_pool_flags"
+          "|supports_omap "
```
ditto

batrick commented on src/mon/OSDMonitor.cc:
```diff
@@ -1995,6 +1995,17 @@ void OSDMonitor::encode_pending(MonitorDBStore::TransactionRef t)
+          dout(10) << __func__ << " replicated pool " << pool_id
+                   << " has OMAP support auto-enabled" << dendl;
+        }
+      }
```
👍

batrick commented on src/mon/OSDMonitor.cc:
```diff
@@ -9125,6 +9149,25 @@ int OSDMonitor::prepare_command_pool_set(const cmdmap_t& cmdmap,
+    if ((val == "true") && p.is_erasure()) {
+      ss << "supports_omap cannot be enabled in ec pools";
+      return -EINVAL;
+    }
```
Unless I'm misunderstanding something, this makes the feature dead code in Umbrella? I thought users would be able to opt-in for their EC pools?

batrick replied:
Furthermore the documentation says:

To enable the new features, the following OSDMap pool settings are required:

- ``allows_ec_overwrites = true`
- ``allows_ec_optimizations = true``
- ``supports_omap = true``

Should this not also check `allows_ec_overwrites` and `allows_ec_optimizations` here as well?

MattyWilliams22 replied:
This check will be changed to `&& !p.allows_ecoptimizations()` in another PR once the omap code has been tested more thoroughly.

EC overwrites are not required to enable omap support, but they are necessary to get some cls methods to work in EC pools, and to use RBD with a single EC pool. So that setting is relevant to the design document but not to this code.

batrick replied:
Ok but:

> Unless I'm misunderstanding something, this makes the feature dead code in Umbrella? I thought users would be able to opt-in for their EC pools?

^ ?

MattyWilliams22 replied:
This feature will be dead code after this PR.

However, we will soon provide a bug fix enabling this feature in Umbrella. We are just working on the last few bug fixes, as we would like all of the teuthology tests to be clean before enabling this feature.

batrick replied:
Hm, I don't particularly like that but ok.

MattyWilliams22 commented:
jenkins test make check

ljflores commented:
@MattyWilliams22 @batrick is this PR ready for a retest?

batrick commented:
> @MattyWilliams22 @batrick is this PR ready for a retest?

It's a funny question. My understanding is that this is tested "in-house" with a patch on top of this branch which addresses 

https://github.com/ceph/ceph/pull/66726#discussion_r3348589214

So it's on them to test it I guess.

I would request that the gate ^ be removed to allow tests to run normally. Add sufficient notice to documentation so users don't turn it on by accident.

batrick commented:
This is the last Umbrella blocker. I would like to branch tomorrow if possible. Please give regular updates.

anthonyeleven commented:
60492 is still in qaOn Jun 4, 2026, at 8:26 AM, Patrick Donnelly ***@***.***> wrote:﻿batrick left a comment (ceph/ceph#66726)
This is the last Umbrella blocker. I would like to branch tomorrow if possible. Please give regular updates.

—Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS and Android. Download it today!
You are receiving this because your review was requested.Message ID: ***@***.***>

MattyWilliams22 commented:
I am currently testing the omap code in teuthology (including the bug fixes which aren't in this PR). 
I have run a set of rados/thrash-erasure-code jobs which have come back clean of any related failures (for the first time).
I have now started a set of rados jobs.

There is a known issue with PeeringState/PGLog which we attempted to fix in this PR (https://github.com/ceph/ceph/pull/68888) but this fix has uncovered other issues that @bill-scales is working on fixing now.
We don't want to enable this feature until that other issue is fixed.

ljflores commented:
@MattyWilliams22 sounds good - if you need it retested by Yuri or Bharath, you can re-add the "needs-qa" label. Feel free to send a ping in the #prs_testing channel in Slack if it needs urgent testing (I added you there).

MattyWilliams22 commented:
Here is a teuthology test run of the rados/thrash-erasure-code directory that is running ceph_test_rados against an EC pool with omap support enabled: https://pulpito-ng.ceph.com/runs/mattywilliams-2026-06-05_08:11:56-rados:thrash-erasure-code-main-distro-default-trial

All 48/48 tests passed.

The branch that was used for this test run included the code from this PR, the omap bug fixes from PR [69267](https://github.com/ceph/ceph/pull/69267), the PeeringState fix from PR [68888](https://github.com/ceph/ceph/pull/68888) and the change to ceph_test_rados from PR [69269](https://github.com/ceph/ceph/pull/69269)

MattyWilliams22 commented:
This force push was to remove a commit that I added. I did not change any other code.

ljflores reviewed (commented):
@MattyWilliams22 @aainscow latest results revealed one issue:

/a/lflores-2026-06-05_19:26:14-rados-wip-lflores-testing-3-2026-06-05-1057-distro-default-trial/
5 jobs: ['269112', '269245', '269313', '269176', '269039']
```
2026-06-05T20:25:58.135 INFO:tasks.workunit.client.0.trial023.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:37: expect_false:  ceph osd pool set pool_getset noscrub asdf
2026-06-05T20:25:58.293 INFO:tasks.workunit.client.0.trial023.stderr:Error EINVAL: expecting value 'true', 'false', '0', or '1'
2026-06-05T20:25:58.297 INFO:tasks.workunit.client.0.trial023.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:37: expect_false:  return 0
2026-06-05T20:25:58.297 INFO:tasks.workunit.client.0.trial023.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:2317: test_mon_osd_pool_set:  expect_false ceph osd pool set pool_getset noscrub 2
2026-06-05T20:25:58.297 INFO:tasks.workunit.client.0.trial023.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:36: expect_false:  set -x
2026-06-05T20:25:58.297 INFO:tasks.workunit.client.0.trial023.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:37: expect_false:  ceph osd pool set pool_getset noscrub 2
2026-06-05T20:25:58.454 INFO:tasks.workunit.client.0.trial023.stderr:Error EINVAL: expecting value 'true', 'false', '0', or '1'
2026-06-05T20:25:58.459 INFO:tasks.workunit.client.0.trial023.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:37: expect_false:  return 0
2026-06-05T20:25:58.460 INFO:tasks.workunit.client.0.trial023.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:2307: test_mon_osd_pool_set:  for flag in nodelete nopgchange nosizechange write_fadvise_dontneed noscrub nodeep-scrub bulk
2026-06-05T20:25:58.460 INFO:tasks.workunit.client.0.trial023.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:2308: test_mon_osd_pool_set:  ceph osd pool set pool_getset nodeep-scrub false
2026-06-05T20:25:58.613 INFO:tasks.workunit.client.0.trial023.stderr:Invalid command: nodeep-scrub not in allow_ec_optimizations|allow_ec_overwrites|bulk|cache_min_evict_age|cache_min_flush_age|cache_target_dirty_high_ratio|cache_target_dirty_ratio|cache_target_full_ratio|compression_algorithm|compression_max_blob_size|compression_min_blob_size|compression_mode|compression_required_ratio|crush_rule|csum_max_block|csum_min_block|csum_type|deep_scrub_interval|dedup_cdc_chunk_size|dedup_chunk_algorithm|dedup_tier|eio|fast_read|fingerprint_algorithm|hashpspool|hit_set_count|hit_set_fpp|hit_set_grade_decay_rate|hit_set_period|hit_set_search_last_n|hit_set_type|min_read_recency_for_promote|min_size|min_write_recency_for_promote|nodedeep-scrub|nodelete|nopgchange|noscrub|nosizechange|pct_update_delay|pg_autoscale_bias|pg_autoscale_mode|pg_num|pg_num_max|pg_num_min|pgp_num|pgp_num_actual|read_ratio|recovery_op_priority|recovery_priority|scrub_max_interval|scrub_min_interval|scrub_priority|set_pool_flags|size|supports_omap|target_max_bytes|target_max_objects|target_size_bytes|target_size_ratio|unset_pool_flags|use_gmt_hitset|write_fadvise_dontneed
2026-06-05T20:25:58.613 INFO:tasks.workunit.client.0.trial023.stderr:osd pool set <pool> <var:allow_ec_optimizations|allow_ec_overwrites|bulk|cache_min_evict_age|cache_min_flush_age|cache_target_dirty_high_ratio|cache_target_dirty_ratio|cache_target_full_ratio|compression_algorithm|compression_max_blob_size|compression_min_blob_size|compression_mode|compression_required_ratio|crush_rule|csum_max_block|csum_min_block|csum_type|deep_scrub_interval|dedup_cdc_chunk_size|dedup_chunk_algorithm|dedup_tier|eio|fast_read|fingerprint_algorithm|hashpspool|hit_set_count|hit_set_fpp|hit_set_grade_decay_rate|hit_set_period|hit_set_search_last_n|hit_set_type|min_read_recency_for_promote|min_size|min_write_recency_for_promote|nodedeep-scrub|nodelete|nopgchange|noscrub|nosizechange|pct_update_delay|pg_autoscale_bias|pg_autoscale_mode|pg_num|pg_num_max|pg_num_min|pgp_num|pgp_num_actual|read_ratio|recovery_op_priority|recovery_priority|scrub_max_interval|scrub_min_interval|scrub_priority|set_pool_flags|size|supports_omap|target_max_bytes|target_max_objects|target_size_bytes|target_size_ratio|unset_pool_flags|use_gmt_hitset|write_fadvise_dontneed> <val> [--yes-i-really-mean-it] :  set pool parameter <var> to <val>
2026-06-05T20:25:58.614 INFO:tasks.workunit.client.0.trial023.stderr:Error EINVAL: invalid command
```
It's coming from [https://github.com/ceph/ceph/pull/66726/changes/b24e907f4b4c81f4e8f4354c83053139fb213460#diff-d2ae0d1f6b17661070357fe[…]7f8eb3b71e0d766e83dcb9c0a137](https://github.com/ceph/ceph/pull/66726/changes/b24e907f4b4c81f4e8f4354c83053139fb213460#diff-d2ae0d1f6b17661070357fefa3bf0383f6cc7f8eb3b71e0d766e83dcb9c0a137). Seems like a simple error to fix. The `nodeep-scrub flag` was accidentally changed to `nodedeep-scrub`.

ljflores commented:
RADOS approved: https://tracker.ceph.com/projects/rados/wiki/TENTACLE#httpstrackercephcomissues77137
